### DISCUSSION TITLE
Bug in building a plan

### DISCUSSION THREAD

**Discussion by BelugaHH** [TUPLES-Trustworthy-AI/Beluga-AI-Challenge#1] [Q&A]
Hi,

I made some tests with your code:

```
json = Path(json_file).read_text()
bpe = BelugaProblemDecoder()
bpr = bpe.decode(json) planner = LazyAstarDeterministicPlanner()
ppl = planner.build_plan(bpr)

```
The last statement caused an error (see below). I used example "problem_63_s105_j3_r2_oc82_f3.json". The same error occurred for the
other examples. I also used my own planner. Always the same issue. Can you provide hints, what is going wrong?

Regards v. turau

[!] Argument jig0001: in atom (empty jig0001: type=typed, empty size=18, loaded size=25, empty=false) is not defined.
[!] Argument type=typed, in atom (empty jig0001: type=typed, empty size=18, loaded size=25, empty=false) is not defined.
[!] Argument empty in atom (empty jig0001: type=typed, empty size=18, loaded size=25, empty=false) is not defined.
[!] Argument size=18, in atom (empty jig0001: type=typed, empty size=18, loaded size=25, empty=false) is not defined.
[!] Argument loaded in atom (empty jig0001: type=typed, empty size=18, loaded size=25, empty=false) is not defined.
[!] Argument size=25, in atom (empty jig0001: type=typed, empty size=18, loaded size=25, empty=false) is not defined.
[!] Argument empty=false in atom (empty jig0001: type=typed, empty size=18, loaded size=25, empty=false) is not defined.
[!] Argument jig0002: in atom (empty jig0002: type=typec, empty size=9, loaded size=18, empty=false) is not defined.
[!] Argument type=typec, in atom (empty jig0002: type=typec, empty size=9, loaded size=18, empty=false) is not defined.
[!] Argument empty in atom (empty jig0002: type=typec, empty size=9, loaded size=18, empty=false) is not defined.
[!] Argument size=9, in atom (empty jig0002: type=typec, empty size=9, loaded size=18, empty=false) is not defined.
[!] Argument loaded in atom (empty jig0002: type=typec, empty size=9, loaded size=18, empty=false) is not defined.
[!] Argument size=18, in atom (empty jig0002: type=typec, empty size=9, loaded size=18, empty=false) is not defined.
[!] Argument empty=false in atom (empty jig0002: type=typec, empty size=9, loaded size=18, empty=false) is not defined.
Traceback (most recent call last):
Argument mismatch in (empty jig0001: type=typed, empty size=18, loaded size=25, empty=false). Predicate expects 1 parameter(s).
Argument mismatch in (empty jig0002: type=typec, empty size=9, loaded size=18, empty=false). Predicate expects 1 parameter(s).
Argument mismatch in (in jig0001: type=typed, empty size=18, loaded size=25, empty=false rack00). Predicate expects 2 parameter(s).
Argument mismatch in (in jig0002: type=typec, empty size=9, loaded size=18, empty=false rack01). Predicate expects 2 parameter(s).
Argument mismatch in (is_type jig0001 typed: size empty=18, size loaded=25). Predicate expects 2 parameter(s).
Argument mismatch in (is_type jig0002 typec: size empty=9, size loaded=18). Predicate expects 2 parameter(s).
Argument mismatch in (is_type jig0003 typed: size empty=18, size loaded=25). Predicate expects 2 parameter(s).
Argument mismatch in (processed-flight beluga1: incoming: none outgoing: typec: size empty=9, size loaded=18). Predicate expects 1 parameter(s).
Argument mismatch in (to_unload jig0003: type=typed, empty size=18, loaded size=25, empty=false beluga3). Predicate expects 2 parameter(s).
Argument mismatch in (to_load typec: size empty=9, size loaded=18 slot0 beluga1). Predicate expects 3 parameter(s).
Argument mismatch in (to_load typed: size empty=18, size loaded=25 slot0 beluga2). Predicate expects 3 parameter(s).

    ppl = planner.build_plan(bpr)
             ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\cygwin/home/ti5vt/tuhh/ift/tuples/beluga/toolkit\evaluation\planner_examples.py", line 139, in build_plan
    domain = SkdPDDLDomain(prb, problem_name='server_side_domain',
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\cygwin/home/ti5vt/tuhh/ift/tuples/beluga/toolkit\skd_domains\skd_pddl_domain.py", line 38,
in __init__ self._create_pddl_structs(domain_str, problem_str, instance_dir)
  File "C:\cygwin/home/ti5vt/tuhh/ift/tuples/beluga/toolkit\skd_domains\skd_base_domain.py", line 326, 
in _create_pddl_structs  self.task: Task = Task(domain, problem)
                                                    ^^^^^^^^^^^^^^^^^^^^^
  File "C:\tools\WPy64-31241\python-3.12.4.amd64\Lib\site-packages\plado\semantics\task.py", line 589,
in __init__ get_state(
  File "C:\tools\WPy64-31241\python-3.12.4.amd64\Lib\site-packages\plado\semantics\task.py", line 498, 
in get_state  args = tuple((obj_ids[a.name] for a in x.arguments))
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\tools\WPy64-31241\python-3.12.4.amd64\Lib\site-packages\plado\semantics\task.py", line 498,
in <genexpr>  args = tuple((obj_ids[a.name] for a in x.arguments))
                                              ~~~~~~~^^^^^^^^
KeyError: 'jig0001:'

**Comment by lompabo** [comment]
Hi,

I've just made a test using the `evaluate_instance.py` script and the lazy A* planner: everything seems to work fine.

Given that you are getting a key error vey early in the evaluation process, your problem might actually be due to issues when decoding the input file. It seems strange, but it might be possible (normally, I always read the problem files using the `json.load` primitive)

Would you mind:

- Trying to run `python evaluate_instances -i <path to the input file> -pln lazy_astar`
- Trying to print the decoded problem `bpe` in your code

Best,

Michele

**Comment by BelugaHH** [comment]
Hi,

I use the following code to generate a solution:
            plan = self.generatePlan()
            plan_json = plan.to_json_str()
            parsed = json.loads(plan_json)
            print(json.dumps(parsed, indent=4))
"generatePlan" is my program and I dump the solution in a file called
solution_63_s105_j3_r2_oc82_f3.json. I have attached the file.
Then I call:
python evaluate_instance.py -i
../benchmarks/scalability_challenge/deterministic/training/problem_63_s105_j3_r2_oc82_f3.json
 --prebuilt-plan ../src/learning/solution_63_s105_j3_r2_oc82_f3.json -pln
lazy_astar

The error message is:
Traceback (most recent call last):
  File
"C:\cygwin\home\ti5vt\tuhh\ift\tuples\beluga\toolkit\evaluate_instance.py",
line 219, in <module>
    evaluator.setup()
  File
"C:\cygwin\home\ti5vt\tuhh\ift\tuples\beluga\toolkit\evaluation\evaluators.py",
line 489, in setup
    self.domain = SkdPDDLDomain(self.prb, self.problem_name,
self.problem_folder, classic=False)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File
"C:\cygwin\home\ti5vt\tuhh\ift\tuples\beluga\toolkit\skd_domains\skd_pddl_domain.py",
line 35, in __init__
    domain_str, problem_str = self._generate_pddl(
                              ^^^^^^^^^^^^^^^^^^^^
  File
"C:\cygwin\home\ti5vt\tuhh\ift\tuples\beluga\toolkit\skd_domains\skd_base_domain.py",
line 373, in _generate_pdd
l
    pddl_problem = encode(
                   ^^^^^^^
  File
"C:\cygwin\home\ti5vt\tuhh\ift\tuples\beluga\toolkit\encoder\pddl_encoding\beluga_pddl_problem_encoding.py",
lin
e 60, in encode
    init_outgoing(variant, beluga_problem, state, problem, domain)
  File
"C:\cygwin\home\ti5vt\tuhh\ift\tuples\beluga\toolkit\encoder\pddl_encoding\beluga_pddl_problem_encoding.py",
lin
e 657, in init_outgoing
    j.name in state.beluga_contents else
              ^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'BelugaProblemState' object has no attribute
'beluga_contents'

Regards
volker

Am Do., 2. Jan. 2025 um 15:47 Uhr schrieb Michele Lombardi <
***@***.***>:
